src/promptflow/promptflow/_core/otel_tracer.py

```diff
@@ -0,0 +1,115 @@
+import json
+import os
+import threading
+from typing import Any, Dict, List, Iterator, Union
+
+from opentelemetry import trace
+from opentelemetry.sdk.trace import Tracer, TracerProvider, Span, StatusCode
+from opentelemetry.sdk.trace.export import (
+    ConsoleSpanExporter,
+    BatchSpanProcessor,
+    SpanExporter,
+    SpanExportResult
+)
+
+from promptflow._utils.dataclass_serializer import serialize
+
+
+# Only for demo purpose, dumping spans to a jsonl file.
+# Console/HTTP/AppInsights exporters are already available.
+# We might build a new Azure Blob exporter leveraging MDC.
+class FileSpanExporter(SpanExporter):
+    def __init__(self, file_name: str):
+        self.file_name = file_name
+        directory = os.path.dirname(file_name)
+        if not os.path.exists(directory):
+            os.makedirs(directory)
+
+    def export(self, spans):
+        with open(self.file_name, "a") as file:
+            for span in spans:
+                file.write(span.to_json(indent=None) + "\n")
+        return SpanExportResult.SUCCESS
+
+
+# Only for demo purpose. This helper class could be part of the telemetry SDK
+class OpenTelemetryTracer:
+    _instance: "OpenTelemetryTracer" = None
+    _lock = threading.Lock()
+
+    @classmethod
+    def get_instance(cls):
```

Comment:
This get_instance should be in context var instead of a singleton? Otherwise if two node are executed in parallel, there would be conflicts.

Reply:
the open telemetry tracer should be singleton and it has it's own context management to handle spans under multi-threading.